File: src/main/java/pl/dmichalski/algorithms/_3_is_anagram/AnagramService1.java
package pl.dmichalski.algorithms._3_is_anagram;

import java.util.Arrays;

class AnagramService1 {

    /**
     * O(n log n) complexity
     */
    public boolean areAnagrams(String text1, String text2) {
        if (text1 == null || text2 == null) {
            return false;
        }
        if (text1.length() == 0 || text2.length() == 0 || text1.length() != text2.length()) {
            return false;
        }

        String text1LowerCase = text1.toLowerCase();
        String text2LowerCase = text2.toLowerCase();

        String text1Sorted = sortString(text1LowerCase);
        String text2Sorted = sortString(text2LowerCase);

        return text1Sorted.equals(text2Sorted);
    }

    private String sortString(String text) {
        char[] characters = text.toCharArray();
        Arrays.sort(characters);
        return new String(characters);
    }

}

File: src/main/kotlin/pl/dmichalski/algorithms/_3_is_anagram/AnagramService1.kt
package pl.dmichalski.algorithms._3_is_anagram

internal class AnagramService1 {

    /**
     * O(n log n) time complexity
     */
    fun areAnagrams(text1: String, text2: String): Boolean {
        if (text1.isEmpty() || text2.isEmpty() || text1.length != text2.length) {
            return false
        }
        val text1LowerCase = text1.toLowerCase()
        val text2LowerCase = text2.toLowerCase()
        val text1Sorted = sortString(text1LowerCase)
        val text2Sorted = sortString(text2LowerCase)
        return text1Sorted == text2Sorted
    }

    private fun sortString(text: String): String {
        return text.toCharArray()
                .sortedArray()
                .concatToString()
    }
}
